base/compiler/tfuncs.jl

```diff
@@ -484,14 +484,27 @@ function typeof_tfunc(@nospecialize(t))
 end
 add_tfunc(typeof, 1, 1, typeof_tfunc, 0)
 
-function typeassert_tfunc(@nospecialize(v), @nospecialize(t))
-    t = instanceof_tfunc(t)[1]
-    t === Any && return v
+function typeassert_instance(@nospecialize(v), @nospecialize(t))
```

Comment:
I'm not too happy about this name, though I don't really have a better idea. Maybe at least drop a comment that `t` here is the one meant to be the "instance" (of a `Type{...}`), not `v`?

Reply:
Maybe just make it an inner function inside `typeassert_tfunc`?